rfd/0163-vnet.md

```diff
@@ -0,0 +1,416 @@
+have logged in to, the VNet domain name will use the address of the leaf proxy.
+For an app named `api` in a Teleport cluster at `leaf.example.com` this
+would look like `api.leaf.example.com.internal`
+This differs from non-VNet app access where both would use the same domain name
+`api.teleport.example.com` because in VNet we cannot use a client certificate or
+a cookie to differentiate.
+
+In the corner case where a user is logged in to the same cluster both as a leaf
+and a root, VNet will prefer to proxy to apps in that cluster as a root.
+
+#### Why `.internal`
+
+Today, by default, apps are publicly reachable at `<app>.<proxy-domain-name>`,
+e.g. `api.teleport.example.com`.
+That is the URL you would go to in the browser to reach an HTTP app through
+Teleport, and it is the same for TCP app access.
+The domain name is expected to resolve to the public address of the Teleport Proxy.
+The app at this address is protected by TLS terminated at the Teleport Proxy,
+which usually has a publicly trusted wildcard TLS certificate for
+`*.teleport.example.com`.
+
+Teleport VNet will avoid overriding or conflicting with this domain name for a
+few reasons:
+
+1. To avoid incidentally breaking the currently supported flow if something goes
+   wrong with the VNet or it is just not running locally.
+1. While the tunnels VNet creates in the background will be authenticated by
+   mTLS, local clients will see it as an unauthenticated TCP endpoint.
+   When users connect to this domain name in the browser, we still want them to
+   see that it is protected by TLS.
+1. HTTP app domain names are sharable and they can be opened by anyone with an
+   account in the same cluster, regardless if they are running VNet or not.
+
+To break the conflict we can add a TLD/suffix to the domain name used in VNet.
+ICANN has recently assigned the `.internal` TLD for internal or private use
+applications so it seems like the appropriate choice.
+https://itp.cdn.icann.org/en/files/root-system/identification-tld-private-use-24-01-2024-en.pdf
+
+### Custom DNS zones
+
+Some users may want to use custom domain names for their apps in VNet.
+Maybe they currently access an app on their VPN at `api.example.private` and
+they want to migrate off their VPN to Teleport without changing all their
+software and scripts that currently reference that name.
+
+To do this they can set the `spec.vnet_addr` field in the app resource to
+`api.example.private`, but VNet will also need to know to configure itself as
+the DNS server for `*.example.private`, and what to do with other DNS requests
+matching `*.example.private` with no matching Teleport app.
+
+To accomodate this the Teleport admin can configure a custom DNS zone
+cluster-wide with a new `vnet` Teleport resource:
+
+```yaml
+version: v1
+kind: vnet
+metadata:
+  name: vnet
+spec:
+  custom_dns_zones:
+  - suffix: .example.private
+    upstream_nameservers:
+    - 1.2.3.4
+    - 4.5.6.7
+  - suffix: .test.private
```

Comment:
It seems unlikely and I don't think it's really a conflict, we can handle both, the performance will just be worse since we'll have to query each cluster to see if the app exists there, and either choose arbitrarily if an app with the same name is in both clusters or reject the request. Shouldn't be that bad though since we'll cache the answer the first time we figure it out